edtlib: Add static type hint stub files for mypy

rruuaanng opened the pull request:
Add stub files (.pyi) for "edtlib.py" and "dtlib.py" to improve mypy static analysis and code completion.

pdgendt commented on scripts/dts/python-devicetree/src/devicetree/dtlib.pyi:
```diff
@@ -0,0 +1,123 @@
+    labels: list[str]
+    parent: Optional[Node]
+    dt: DT
+    def __init__(self, name: str, parent: Node | None, dt: DT, filename: str, lineno: int) -> None: ...
```
This sadly isn't picked up by the `ruff formatter`, but it would be nice to have consistent maximum 100 line lengths.

pdgendt replied:
Could you run the following to make it fully consisten?

```shell
 $ ruff format scripts/dts/python-devicetree/src/devicetree/*.pyi
```

rruuaanng replied:
To be honest, I formatted them manually, as `ruff format` still seems to retain some longer segments.

pdgendt replied:
It should pick up the rules from `.ruff.toml`?

Ideally both the formatter and the check are verified, I've also created a PR to check `.pyi` files in CI: https://github.com/zephyrproject-rtos/zephyr/pull/88025

rruuaanng replied:
It looks good. I will also add a stub for `class GpioReservedRange` mentioned in #88023 in the future.

rruuaanng commented:
Let me restart the ci process.

rruuaanng commented:
@mbolivar @pdgendt Please review it again!

pdgendt commented:
> Please review it again!

Hmn, with the `ruff` checks now added, I think it would be best to fix the following issues too:

```toml
# in .ruff-excludes.toml
"./scripts/dts/python-devicetree/src/devicetree/dtlib.py" = [
    "UP007",    # https://docs.astral.sh/ruff/rules/non-pep604-annotation
]
"./scripts/dts/python-devicetree/src/devicetree/edtlib.py" = [
    "UP007",    # https://docs.astral.sh/ruff/rules/non-pep604-annotation
]
```

So that the interface and actual implementation have the same declarations.

kartben reviewed (commented):
It feels like some documentation needs to be added as to how/when these files need to be regenerated?

henrikbrixandersen commented:
@rruuaanng Please rebase on `main` and resolve any merge conflicts.

rruuaanng commented:
@henrikbrixandersen 
I'll deal with it after the vacation :)

mbolivar commented:
@rruuaanng can you please check CI and update again so we can get this in?

kartben commented:
> It feels like some documentation needs to be added as to how/when these files need to be regenerated?

@rruuaanng I think this never got addressed? How/when do you expect people to update these files? There's already good do's/don'ts at the top of dtlib/edtlib files so it would probably be good to add a not there w.r.t what people must do what they add new APIs etc

mbolivar commented:
> > > It feels like some documentation needs to be added as to how/when these files need to be regenerated?
> > 
> > 
> > @rruuaanng I think this never got addressed? How/when do you expect people to update these files? There's already good do's/don'ts at the top of dtlib/edtlib files so it would probably be good to add a not there w.r.t what people must do what they add new APIs etc
> 
> Maybe I should mention that in the README. I'll add it.

In my experience, nobody reads the README -- you might prefer to add a comment at the top of the source file as well.

mbolivar commented:
> If developers use `stubgen` command to generate these stubs, the comments will be lost, perhaps a `README.md` would be more appropriate. WDYT?

Feel free to do it later, I think it's OK as-is

ljd42 reviewed (changes requested):
Hi @rruuaanng ,

On overall looks good. Possibly some oversights, in addition to some missing prototype (were this left out intentionally?).

Thx!

rruuaanng committed:
edtlib: Add static type hint stub files for mypy

Add stub files (.pyi) for "edtlib.py" and "dtlib.py" to
improve mypy static analysis and code completion.

Signed-off-by: James Roy <rruuaanng@outlook.com
